File: pledit/Gui.py
''' Gui.py

GUI for pledit.
'''

# Copyright (C) 2005 W. Evan Sheehan
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

import pygtk, gtk, gtk.glade

class Main:
    def __init__ (self):
        window = gtk.glade.XML ("pledit.glade")
        playlist = window.get_widget ("playlist")

        # Create a ListStore for the playlist and add it to the TreeView.
        self.songList = gtk.ListStore (gobject.TYPE_STRING,
                                       gobject.TYPE_INT,
                                       gobject.TYPE_STRING)
        playlist.set_model (self.songList)

        # Make a column that displays the desired track info for the playlist.
        cell = gtk.CellRendererText ()
        cell.mode = gtk.CELL_RENDERER_MODE_EDITABLE
        col = gtk.TreeViewColumn ("Track", cell, text=0)
        playlist.append_column (col)

        # Connect to all our signals.
        window.signal_autoconnect (self)

    def _RemoveClicked (self, button, data=None):
        ''' Remove button callback. '''
        print "remove"

    def _AddClicked (self, button, data=None):
        ''' Add button callback. '''
        SongChooser (self.songList)

    def _NewActivated (self, item, data=None):
        ''' New menu item clicked. '''
        pass

    def _OpenActivate (self, item, data=None):
        ''' Open menu item clicked. '''
        pass

    def _SaveActivated (self, item, data=None):
        ''' Save menu item clicked. '''
        pass

    def _SaveAsActivated (self, item, data=None):
        ''' Save As menu item clicked. '''
        pass

    def _QuitActivated (self, item, data=None):
        ''' Quit menu item clicked. '''
        # FIXME Check if the playlist has been saved, if not prompt.
        gtk.main_quit ()

    def _CutActivated (self, item, data=None):
        ''' Cut menu item clicked. '''
        pass

    def _CopyActivated (self, item, data=None):
        ''' Copy menu item clicked. '''
        pass

    def _PasteActivated (self, item, data=None):
        ''' Paste menu item clicked. '''
        pass

    def _DeleteActivated (self, item, data=None):
        ''' Delete menu item clicked. '''
        pass

    def _AboutActivated (self, item, data=None):
        ''' About menu item clicked. '''
        pass

    def _Delete (self, widget, event, data=None):
        ''' Bye bye. '''
        gtk.main_quit ()


class SongChooser:
    def __init__ (self, list):
        pass


class AboutBox:
    def __init__ (self):
        pass


### The End ###
# vim:ts=4:sw=4:et:tw=80

File: screamers/src/Gui.cpp
/*
 * Gui.cpp - various GUI screens
 *
 * Copyright (C) 2005 Screamers Group (see AUTHORS)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 */

#include "Gui.h"

namespace Screamers
{
namespace Client
{

static CEGUI::MouseButton convertOgreButtonToCegui (int button)
{
	switch (button) {
	case Ogre::MouseEvent::BUTTON0_MASK:	return CEGUI::LeftButton;
	case Ogre::MouseEvent::BUTTON1_MASK:	return CEGUI::RightButton;
	case Ogre::MouseEvent::BUTTON2_MASK:	return CEGUI::MiddleButton;
	case Ogre::MouseEvent::BUTTON3_MASK:	return CEGUI::X1Button;
	default:				return CEGUI::LeftButton;
	}
}

GuiFrameListener::GuiFrameListener (Ogre::RenderWindow *window, CEGUI::Renderer *renderer) : FrameListener (window)
{
	event_processor = new Ogre::EventProcessor ();
	event_processor->initialise (window);
	event_processor->startProcessingEvents ();
	event_processor->addKeyListener (this);
	event_processor->addMouseMotionListener (this);
	event_processor->addMouseListener (this);

	gui_renderer = renderer;

	quit = false;
}

bool GuiFrameListener::frameEnded (const Ogre::FrameEvent &event)
{
	if (quit)
		return false;
	return FrameListener::frameEnded (event);
}

void GuiFrameListener::mouseMoved (Ogre::MouseEvent *event)
{
	CEGUI::System::getSingleton ().injectMouseMove (event->getRelX () * gui_renderer->getWidth (),
							event->getRelY () * gui_renderer->getHeight ());
}

void GuiFrameListener::mouseDragged (Ogre::MouseEvent *event)
{
	mouseMoved (event);
}

void GuiFrameListener::mousePressed (Ogre::MouseEvent *event)
{
	CEGUI::System::getSingleton ().injectMouseButtonDown (convertOgreButtonToCegui (event->getButtonID ()));
}

void GuiFrameListener::mouseReleased (Ogre::MouseEvent *event)
{
	CEGUI::System::getSingleton ().injectMouseButtonUp (convertOgreButtonToCegui (event->getButtonID ()));
}

void GuiFrameListener::keyPressed (Ogre::KeyEvent *event)
{
	// if the user hit escape, quit. if not, punt back into CEGUI
	if (event->getKey () == Ogre::KC_ESCAPE) {
		quit = true;
	} else {
		CEGUI::System::getSingleton ().injectKeyDown (event->getKey ());
		CEGUI::System::getSingleton ().injectChar (event->getKeyChar ());
	}
	event->consume ();
}

void GuiFrameListener::keyReleased (Ogre::KeyEvent *event)
{
	CEGUI::System::getSingleton ().injectKeyUp (event->getKey ());
	event->consume ();
}

void GuiFrameListener::keyClicked (Ogre::KeyEvent *event)
{
	event->consume ();
}

Gui::Gui (Ogre::RenderWindow *window)
{
	this->window = window;
	gui_renderer = new CEGUI::OgreCEGUIRenderer(window, Ogre::RENDER_QUEUE_OVERLAY, false, 0, Ogre::ST_GENERIC);
	gui_system = new CEGUI::System (gui_renderer);
}

Gui::~Gui ()
{
	if (gui_system) {
		delete gui_system;
		gui_system = NULL;
	}
	if (gui_renderer) {
		delete gui_renderer;
		gui_renderer = NULL;
	}
}

};
};
